# Exploring eBay Car Sales Data

## About

This project is about a dataset of used cars from eBay Kleinanzeigen, a [classifieds](https://en.wikipedia.org/wiki/Classified_advertising) section of the German eBay website. The dataset was originally [scraped](https://en.wikipedia.org/wiki/Web_scraping) and uploaded to Kaggle by user [orgesleka](https://www.kaggle.com/orgesleka).
The original dataset isn't available on Kaggle anymore, but you can find it [here](https://data.world/data-society/used-cars-data).

## Note
A few modifications from the original dataset:

- We sampled 50,000 data points from the full dataset, to ensure your code runs quickly in our hosted environment
- We dirtied the dataset a bit to more closely resemble what you would expect from a scraped dataset (the version uploaded to Kaggle was cleaned to be easier to work with)


## Goal
The aim of this project is to clean the data and analyze the included used car listings.

In [1]:
# import pandas and numpy
import pandas as pd
import numpy as np

#i got an encoding error, 
#so i tried to specify two most popular encodings (Latin-1 and Windows-1252)
autos = pd.read_csv("autos.csv", encoding="Windows-1252") 

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

- The dataset contains 20 columns, most of them are strings.
- Some columns (vehicleType, gearbox, model, fuelType, notRepairedDamage) have null values, but none have more than ~20% null values.

In [4]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores.

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
# following edits to column names (from camelcase to snakecase)
col_edit = {
    'yearOfRegistration':'registration_year', 
    'monthOfRegistration':'registration_month', 
    'notRepairedDamage':'unrepaired_damage', 
    'dateCreated':'ad_created', 
    'offerType':'offer_type', 
    'vehicleType':'vehicle_type', 
    'powerPS':'power_ps', 
    'fuelType':'fuel_type', 
    'nrOfPictures':'nr_of_pictures', 
    'postalCode':'postal_code', 
    'lastSeen':'last_seen', 
    'dateCrawled':'date_crawled'
}

#the modified columns names are assigned back to the dataframe
autos.rename(col_edit, axis=1, inplace=True) 


In [7]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [8]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [9]:
# to look at descriptive statistics for all columns
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-21 20:37:19  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

In [10]:
#candidates to be dropped (almost all the values are the same)
autos["seller"].value_counts() 

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [11]:
#candidates to be dropped (almost all the values are the same)
autos["offer_type"].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

In [12]:
autos["abtest"].value_counts()

test       25756
control    24244
Name: abtest, dtype: int64

In [13]:
autos["vehicle_type"].value_counts()

limousine     12859
kleinwagen    10822
kombi          9127
bus            4093
cabrio         3061
coupe          2537
suv            1986
andere          420
Name: vehicle_type, dtype: int64

In [14]:
autos["gearbox"].value_counts() #needs to be cleaned (from german to english)

manuell      36993
automatik    10327
Name: gearbox, dtype: int64

In [15]:
autos["odometer"].value_counts()#needs to be cleaned (remove the "km" and ",") and convert to integer

150,000km    32424
125,000km     5170
100,000km     2169
90,000km      1757
80,000km      1436
70,000km      1230
60,000km      1164
50,000km      1027
5,000km        967
40,000km       819
30,000km       789
20,000km       784
10,000km       264
Name: odometer, dtype: int64

In [16]:
autos["fuel_type"].value_counts() #needs to be cleaned (from german to english)

benzin     30107
diesel     14567
lpg          691
cng           75
hybrid        37
andere        22
elektro       19
Name: fuel_type, dtype: int64

In [17]:
autos["unrepaired_damage"].value_counts() #needs to be cleaned (from german to english)

nein    35232
ja       4939
Name: unrepaired_damage, dtype: int64

It looks like that the price and odometer columns are numeric values stored as text. For each column:

- Remove any non-numeric characters.
- Convert the column to a numeric dtype.
- Use DataFrame.rename() to rename the column to odometer_km.


## Converting price and odometer columns to numeric values

In [18]:
autos["price"] = autos["price"].str.replace("$", "").str.replace(",", "").astype(int)
autos["odometer"] = autos["odometer"].str.replace("km", "").str.replace(",", "").astype(int)

In [19]:
autos.rename({"odometer":"odometer_km"}, axis = 1, inplace = True)

In [20]:
print(autos["price"].dtype)
print(autos["odometer_km"].dtype)

int64
int64


In [21]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price   abtest vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   

     gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere       150000                   3       lpg   
1  automatik       286     7er       150000                   6    benzin   
2    manuell       102    golf        70000                   7    benzin   
3  automatik        71  fortwo        70000                   6    benzin   
4    manuell         0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

## Drop the columns (outliers)

In [22]:
drop_columns = ["seller", "nr_of_pictures", "offer_type"]
autos.drop(drop_columns, axis = 1, inplace = True)

## Calculate the distribution of values date_crawled, ad_created, and last_seen

In [23]:
# data crawled
print('The counts of the column data_crawled are:')
autos['date_crawled'].str[:10].value_counts(normalize=True, 
                                            dropna=False).sort_index()

The counts of the column data_crawled are:


2016-03-05    0.02538
2016-03-06    0.01394
2016-03-07    0.03596
2016-03-08    0.03330
2016-03-09    0.03322
2016-03-10    0.03212
2016-03-11    0.03248
2016-03-12    0.03678
2016-03-13    0.01556
2016-03-14    0.03662
2016-03-15    0.03398
2016-03-16    0.02950
2016-03-17    0.03152
2016-03-18    0.01306
2016-03-19    0.03490
2016-03-20    0.03782
2016-03-21    0.03752
2016-03-22    0.03294
2016-03-23    0.03238
2016-03-24    0.02910
2016-03-25    0.03174
2016-03-26    0.03248
2016-03-27    0.03104
2016-03-28    0.03484
2016-03-29    0.03418
2016-03-30    0.03362
2016-03-31    0.03192
2016-04-01    0.03380
2016-04-02    0.03540
2016-04-03    0.03868
2016-04-04    0.03652
2016-04-05    0.01310
2016-04-06    0.00318
2016-04-07    0.00142
Name: date_crawled, dtype: float64

In [24]:
# ad_created
print('The counts of the column ad_created are:')
autos['ad_created'].str[:10].value_counts(normalize=True, 
                                            dropna=False).sort_index()

The counts of the column ad_created are:


2015-06-11    0.00002
2015-08-10    0.00002
2015-09-09    0.00002
2015-11-10    0.00002
2015-12-05    0.00002
               ...   
2016-04-03    0.03892
2016-04-04    0.03688
2016-04-05    0.01184
2016-04-06    0.00326
2016-04-07    0.00128
Name: ad_created, Length: 76, dtype: float64

In [25]:
# last_seen
print('The counts of the column last_seen are:')
autos['last_seen'].str[:10].value_counts(normalize=True, 
                                            dropna=False).sort_index()

The counts of the column last_seen are:


2016-03-05    0.00108
2016-03-06    0.00442
2016-03-07    0.00536
2016-03-08    0.00760
2016-03-09    0.00986
2016-03-10    0.01076
2016-03-11    0.01252
2016-03-12    0.02382
2016-03-13    0.00898
2016-03-14    0.01280
2016-03-15    0.01588
2016-03-16    0.01644
2016-03-17    0.02792
2016-03-18    0.00742
2016-03-19    0.01574
2016-03-20    0.02070
2016-03-21    0.02074
2016-03-22    0.02158
2016-03-23    0.01858
2016-03-24    0.01956
2016-03-25    0.01920
2016-03-26    0.01696
2016-03-27    0.01602
2016-03-28    0.02086
2016-03-29    0.02234
2016-03-30    0.02484
2016-03-31    0.02384
2016-04-01    0.02310
2016-04-02    0.02490
2016-04-03    0.02536
2016-04-04    0.02462
2016-04-05    0.12428
2016-04-06    0.22100
2016-04-07    0.13092
Name: last_seen, dtype: float64

In [26]:
import datetime as dt

cols = ["date_crawled", "ad_created", "last_seen"]

for c in cols:
    print("The counts of the column {} are".format(c), "\n")  
    min_val = dt.datetime.strptime(autos[c].str[:10].min(), "%Y-%m-%d")
    max_val = dt.datetime.strptime(autos[c].str[:10].max(), "%Y-%m-%d")
    print("The min value of the column {} is {}".format(c, min_val.strftime("%Y-%m-%d")))
    print("The max value of the column {} is {}".format(c, max_val.strftime("%Y-%m-%d")))
    print("The difference in days are {}".format(str(max_val - min_val)[:7]))
    print("\n")

The counts of the column date_crawled are 

The min value of the column date_crawled is 2016-03-05
The max value of the column date_crawled is 2016-04-07
The difference in days are 33 days


The counts of the column ad_created are 

The min value of the column ad_created is 2015-06-11
The max value of the column ad_created is 2016-04-07
The difference in days are 301 day


The counts of the column last_seen are 

The min value of the column last_seen is 2016-03-05
The max value of the column last_seen is 2016-04-07
The difference in days are 33 days




Insights from the datecrawled, ad_created and lastseen columns are:

- The data crawling was done from 5th march till 7th april 2016 (33 days)
- The ads are created from almost 10 months (301 days)
- The last seen is also at the same timeline with data crawling when it was done. (from 5th march till 7th april 2016 for 33 days)



## Calculate the distribution of registration_year

In [27]:
autos["registration_year"].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Min and max value appear to be off and odd from the range in the registration_year. 
- Min value is 1000, many years in the past before cars even were invented. 
- Max value is 9999, many years in the future.

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

In [28]:
autos["registration_year"].value_counts(dropna = False).sort_index(ascending = True).head(50)

1000     1
1001     1
1111     1
1500     1
1800     2
1910     9
1927     1
1929     1
1931     1
1934     2
1937     4
1938     1
1939     1
1941     2
1943     1
1948     1
1950     3
1951     2
1952     1
1953     1
1954     2
1955     2
1956     5
1957     2
1958     4
1959     7
1960    34
1961     6
1962     4
1963     9
1964    12
1965    17
1966    22
1967    27
1968    26
1969    19
1970    45
1971    27
1972    35
1973    26
1974    24
1975    19
1976    27
1977    22
1978    47
1979    35
1980    97
1981    31
1982    43
1983    53
Name: registration_year, dtype: int64

In [29]:
autos["registration_year"].value_counts(dropna = False).sort_index(ascending = True).tail(20)

2011    1634
2012    1323
2013     806
2014     666
2015     399
2016    1316
2017    1453
2018     492
2019       3
2800       1
4100       1
4500       1
4800       1
5000       4
5911       1
6200       1
8888       1
9000       2
9996       1
9999       4
Name: registration_year, dtype: int64

In [30]:
# Because the last ad created was in 2016, 
# only max register year until 2016 is considered
autos = autos[autos["registration_year"].between(1954, 2016)]

In [31]:
autos["registration_year"].value_counts(dropna = False).sort_index(ascending = True).tail(20)

1997    2028
1998    2453
1999    3000
2000    3354
2001    2703
2002    2533
2003    2727
2004    2737
2005    3015
2006    2708
2007    2304
2008    2231
2009    2098
2010    1597
2011    1634
2012    1323
2013     806
2014     666
2015     399
2016    1316
Name: registration_year, dtype: int64

In [32]:
autos["registration_year"].value_counts(normalize=True, dropna=False).sort_index(ascending=True).tail(20)

1997    0.042253
1998    0.051107
1999    0.062504
2000    0.069879
2001    0.056316
2002    0.052774
2003    0.056816
2004    0.057024
2005    0.062816
2006    0.056420
2007    0.048003
2008    0.046482
2009    0.043711
2010    0.033273
2011    0.034044
2012    0.027564
2013    0.016793
2014    0.013876
2015    0.008313
2016    0.027418
Name: registration_year, dtype: float64

## Explore the brand

In [33]:
autos["brand"].describe()

count          47997
unique            40
top       volkswagen
freq           10187
Name: brand, dtype: object

In [34]:
autos["brand"].value_counts(dropna = False, normalize = True)

volkswagen        0.212242
bmw               0.110069
opel              0.108173
mercedes_benz     0.095360
audi              0.086443
ford              0.069775
renault           0.047357
peugeot           0.029544
fiat              0.025877
seat              0.018168
skoda             0.016043
mazda             0.015147
nissan            0.015105
citroen           0.013938
smart             0.013918
toyota            0.012480
sonstige_autos    0.010667
hyundai           0.009855
volvo             0.009251
mini              0.008646
mitsubishi        0.008146
honda             0.007855
kia               0.007105
alfa_romeo        0.006605
porsche           0.006084
suzuki            0.005917
chevrolet         0.005709
chrysler          0.003667
dacia             0.002563
daihatsu          0.002563
jeep              0.002250
subaru            0.002188
land_rover        0.002021
saab              0.001604
jaguar            0.001583
trabant           0.001542
daewoo            0.001500
r

Lets explore top 6 brands that have certain percentage of the total values > 5%

In [35]:
brands_top_6 = autos["brand"].value_counts(dropna = False, normalize = True).index[:6]
brands_top_6

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')

In [50]:
brand_price = {}

for brand in brands_top_6:
    mean_price = autos[autos["brand"] == brand]["price"].mean()
    brand_price[brand] = mean_price

brand_price

{'volkswagen': 6512.984195543339,
 'bmw': 8334.045996592846,
 'opel': 5254.904275808937,
 'mercedes_benz': 30304.17544242954,
 'audi': 9093.65003615329,
 'ford': 7254.293520453867}

The number 1 brand volkswagen has a mean price of $6512 and it might be its affordability and not being a luxury brand, there are lot of ads for this brand of ebay.

Overall, in the top 6 brands, there's a distinct price gap.
- Audi, BMW and Mercedes Benz are more expensive
- Ford and Opel are less expensive
- Volkswagen is in between


In [51]:
bmp_series = pd.Series(brand_price)
bmp_series

volkswagen        6512.984196
bmw               8334.045997
opel              5254.904276
mercedes_benz    30304.175442
audi              9093.650036
ford              7254.293520
dtype: float64

In [52]:
df = pd.DataFrame(bmp_series, columns=['mean_price'])
df

mean_price
volkswagen      6512.984196
bmw             8334.045997
opel            5254.904276
mercedes_benz  30304.175442
audi            9093.650036
ford            7254.293520

In [53]:
brand_mileage = {}

for brand in brands_top_6:
    mean_mileage = autos[autos["brand"] == brand]["odometer_km"].mean()
    brand_mileage[brand] = mean_mileage

brand_mileage

{'volkswagen': 128733.18935898694,
 'bmw': 132458.8302101079,
 'opel': 129252.69645608628,
 'mercedes_benz': 130927.4634039764,
 'audi': 129287.78018799711,
 'ford': 124151.98566736339}

In [55]:
df["mean_mileage"] = pd.Series(brand_mileage)
df

mean_price   mean_mileage
volkswagen      6512.984196  128733.189359
bmw             8334.045997  132458.830210
opel            5254.904276  129252.696456
mercedes_benz  30304.175442  130927.463404
audi            9093.650036  129287.780188
ford            7254.293520  124151.985667

All the brands seems to have almost same mileage, which points to affordability as we mentioned above might be the reason for high popularity on ebay.

## Price to Mileage Relation

In [56]:
print(sorted(autos["odometer_km"].unique()))

[5000, 10000, 20000, 30000, 40000, 50000, 60000, 70000, 80000, 90000, 100000, 125000, 150000]


In [57]:
# create a list of our conditions
conditions = [
    (autos['odometer_km'] <= 30000),
    (autos['odometer_km'] > 30000) & (autos['odometer_km'] <= 60000),
    (autos['odometer_km'] > 60000) & (autos['odometer_km'] <= 90000),
    (autos['odometer_km'] > 90000)
    ]

# create a list of the values we want to assign for each condition
values = ['low', 'avg', 'high', 'very_high']

# create a new column and use np.select to assign values to it using our lists as arguments
autos['mileage_category'] = np.select(conditions, values)

# display updated DataFrame
autos.head()

<ipython-input-57-c202904cca56>:13: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  autos['mileage_category'] = np.select(conditions, values)


date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price   abtest vehicle_type  registration_year    gearbox  power_ps  \
0   5000  control          bus               2004    manuell       158   
1   8500  control    limousine               1997  automatik       286   
2   8990     test    limousine               2009    manuell       102   
3   4350  control   kleinwagen               2007  automatik        71   
4   1350     test        kombi               2003    manuell         0   

    model  odometer_km  registration_month fuel_type       brand  \
0  andere       150000                   3       lpg     peugeot   
1     7er       150000                   6    benzin         bmw   
2    golf        70000                   7    benzin  volkswagen   
3  fortwo        70000                   6    benzin       smart   
4   focus       150000                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  \
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54   
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08   
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37   
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28   
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50   

  mileage_category  
0        very_high  
1        very_high  
2             high  
3             high  
4        very_high

In [58]:
autos["mileage_category"].value_counts(dropna = False, normalize = True)

very_high    0.793654
high         0.089381
avg          0.061254
low          0.055712
Name: mileage_category, dtype: float64

~79% of the cars in the dataset belong to very_high mileage category with mileage over 100,000 km.

In [60]:
mileage_cat_unique = autos["mileage_category"].unique()
mileage_cat_unique

array(['very_high', 'high', 'avg', 'low'], dtype=object)

In [61]:
mileage_price = {}

for cat in mileage_cat_unique:
    mean_price = autos[autos["mileage_category"] == cat]["price"].mean()
    mileage_price[cat] = mean_price

mileage_price 

{'very_high': 7820.595332475783,
 'high': 9439.913986013986,
 'avg': 27096.449659863945,
 'low': 15663.271503365744}

From the above, we can observe that the price of the cars with negatively correlated to the mileage. i.e. If the mileage is high, the price is low, and if the mileage is low, the price seems to be high.

## Most common brand/model

In [62]:
brand_model = autos["brand"] + "/" + autos["model"] 
brand_model.value_counts(dropna = False, normalize = True)

volkswagen/golf     0.079484
bmw/3er             0.056004
NaN                 0.051087
volkswagen/polo     0.034940
opel/corsa          0.034252
                      ...   
audi/200            0.000021
rover/rangerover    0.000021
bmw/i3              0.000021
rover/discovery     0.000021
ford/b_max          0.000021
Length: 291, dtype: float64

Volkswagen Golf is the most common car which are around 8% of the dataset.

## Converting field names of gearbox, unrepaired_damage and fueltype from german to English

In [65]:
autos["gearbox"].unique()

array(['manuell', 'automatik', nan], dtype=object)

In [66]:
gearbox_dict = {
    'manuell':'manual',
    'automatik':'automatic',
    np.nan:'unknown'
}

autos["gearbox"] = autos["gearbox"].map(gearbox_dict)
autos["gearbox"].value_counts(dropna = False)

<ipython-input-66-a1586c3fa46a>:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  autos["gearbox"] = autos["gearbox"].map(gearbox_dict)


manual       35543
automatic    10043
unknown       2411
Name: gearbox, dtype: int64

In [67]:
autos["fuel_type"].unique()

array(['lpg', 'benzin', 'diesel', nan, 'cng', 'hybrid', 'elektro',
       'andere'], dtype=object)

In [68]:
fueltype_dict = {
    'lpg' : 'lpg',
    'benzin' : 'petrol',
     np.nan : 'unknown',
    'cng' : 'cng',
    'hybrid' : 'hybrid',
    'elektro' : 'electric',
    'andere' : 'other'
}

autos["fuel_type"] = autos["fuel_type"].map(fueltype_dict)
autos["fuel_type"].value_counts(dropna = False)

<ipython-input-68-fb3cd98578d1>:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  autos["fuel_type"] = autos["fuel_type"].map(fueltype_dict)


petrol      29229
NaN         14233
unknown      3717
lpg           671
cng            72
hybrid         37
electric       19
other          19
Name: fuel_type, dtype: int64

In [69]:
autos["unrepaired_damage"].unique()

array(['nein', nan, 'ja'], dtype=object)

In [70]:
unrepaired_dict = {
    'nein' : 'no',
    'ja' : 'yes',
    np.nan : 'unknown'
}

autos["unrepaired_damage"] = autos["unrepaired_damage"].map(unrepaired_dict)
autos["unrepaired_damage"].value_counts(dropna = False)

<ipython-input-70-e242de41e87e>:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  autos["unrepaired_damage"] = autos["unrepaired_damage"].map(unrepaired_dict)


no         34246
unknown     8972
yes         4779
Name: unrepaired_damage, dtype: int64

## Exploring Odometer_km column

In [71]:
autos["odometer_km"].unique().shape[0]  #No.of unique values

13

In [72]:
autos["odometer_km"].describe()

count     47997.000000
mean     125598.995771
std       40042.930768
min        5000.000000
25%      100000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [73]:
autos["odometer_km"].value_counts(dropna = False)

150000    31026
125000     4960
100000     2107
90000      1695
80000      1396
70000      1199
60000      1134
50000      1007
5000        895
40000       799
30000       769
20000       762
10000       248
Name: odometer_km, dtype: int64

In [74]:
autos["odometer_km"].value_counts().sort_index(ascending = False).head()

150000    31026
125000     4960
100000     2107
90000      1695
80000      1396
Name: odometer_km, dtype: int64

The odometer doesn't seem to have any outliers to be concerning.

## Exploring price column

In [75]:
autos["price"].unique().shape[0]  #No.of unique values

2334

In [76]:
autos["price"].describe()

count    4.799700e+04
mean     9.582981e+03
std      4.845378e+05
min      0.000000e+00
25%      1.150000e+03
50%      2.990000e+03
75%      7.400000e+03
max      1.000000e+08
Name: price, dtype: float64

In [77]:
autos["price"].value_counts(dropna = False)

0        1331
500       756
1500      696
2500      614
1200      606
         ... 
2671        1
27299       1
4780        1
686         1
84997       1
Name: price, Length: 2334, dtype: int64

In [78]:
autos["price"].value_counts(dropna = False).sort_index(ascending = True).tail(20)

197000      1
198000      1
220000      1
250000      1
259000      1
265000      1
295000      1
299000      1
345000      1
350000      1
999990      1
999999      2
1234566     1
1300000     1
3890000     1
10000000    1
11111111    2
12345678    1
27322222    1
99999999    1
Name: price, dtype: int64

The price seems to have outliers at the higher end. So we will just proceed with the price ranges from 1000 to 100000

In [79]:
autos = autos[autos["price"].between(1000, 100000)]
autos["price"].value_counts(dropna = False).sort_index(ascending = True).head(20)

1000    602
1039      1
1040      1
1049      6
1050     89
1059      1
1070      1
1080      6
1090      4
1095      2
1098      1
1099     41
1100    355
1111     38
1112      1
1119      1
1120      2
1149     10
1150    212
1169      1
Name: price, dtype: int64

### How much cheaper are the damaged cars than their non damaged counterparts?

In [81]:
damage_unique = (autos[autos["unrepaired_damage"].isin(["yes", "no"])]["unrepaired_damage"]
                                                                            .value_counts()
                                                                            .index
                )
damage_unique

Index(['no', 'yes'], dtype='object')

In [82]:
is_damage_price = {}

for var in damage_unique:
    mean_price = autos[autos["unrepaired_damage"] == var]["price"].mean()
    is_damage_price[var] = mean_price

is_damage_price 

{'no': 7861.325567281432, 'yes': 3866.8081244598097}

In [109]:
price_diff = round((is_damage_price["yes"] - is_damage_price["no"])/is_damage_price["no"],2)
price_diff

-0.51

The cars that are damaged are ~51% cheaper than their non-damaged counterparts.